service/resource/resourcegroup/resource.go

```diff
@@ -0,0 +1,289 @@
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	tags := map[string]string{
+		clusterIDTag:  key.ClusterID(customObject),
+		customerIDTag: key.ClusterCustomer(customObject),
+	}
+	resourceGroup := &Group{
+		Name:     key.ClusterID(customObject),
+		Location: key.Location(customObject),
+		Tags:     tags,
+	}
+
+	return resourceGroup, nil
+}
+
+// GetCreateState returns the resource group for this cluster if it should be
+// created.
+func (r *Resource) GetCreateState(obj, currentState, desiredState interface{}) (interface{}, error) {
+	currentResourceGroup, err := toGroup(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredResourceGroup, err := toGroup(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var resourceGroupToCreate *Group
+	if currentResourceGroup == nil || currentResourceGroup.Name == "" {
+		resourceGroupToCreate = desiredResourceGroup
+	}
+
+	return resourceGroupToCreate, nil
+}
+
+// GetDeleteState returns the resource group for this cluster if it should be
+// deleted.
+func (r *Resource) GetDeleteState(obj, currentState, desiredState interface{}) (interface{}, error) {
+	currentResourceGroup, err := toGroup(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredResourceGroup, err := toGroup(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var resourceGroupToDelete *Group
+	if currentResourceGroup != nil && currentResourceGroup.Name != "" {
+		resourceGroupToDelete = desiredResourceGroup
+	}
+
+	return resourceGroupToDelete, nil
+}
+
+// GetUpdateState returns an empty group for the create, delete and update
+// states because resource groups are not updated.
+func (r *Resource) GetUpdateState(obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
+	return &Group{}, &Group{}, &Group{}, nil
+}
+
+// Name returns the resource name.
+func (r *Resource) Name() string {
+	return Name
+}
+
+// ProcessCreateState creates the resource group via the Azure API.
+func (r *Resource) ProcessCreateState(obj, createState interface{}) error {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	resourceGroupToCreate, err := toGroup(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if resourceGroupToCreate != nil && resourceGroupToCreate.Name != "" {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "creating the resource group in the Azure API")
+
+		groupClient, err := r.getGroupsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		resourceGroup := azureresource.Group{
+			Name:      to.StringPtr(resourceGroupToCreate.Name),
+			Location:  to.StringPtr(resourceGroupToCreate.Location),
+			ManagedBy: to.StringPtr(managedBy),
+			Tags:      to.StringMapPtr(resourceGroupToCreate.Tags),
```

Comment:
Why do we pass around custom `Group` object instead of using `azureresource.Group`?

Reply:
I defined a separate struct because we only care about a subset of the Azure SDK object.

To me this felt cleaner but it does add some overhead as we need to copy the values between the SDK object and our struct.